### PR TITLE
usm: Refactor istio monitor to use new uprobe attacher

### PR DESCRIPTION
**Pull request by gjulianm**
### What does this PR do?

This PR refactors the istio monitor to use the new uprobe attacher. 

### Motivation

Use the common infrastructure for uprobe attachment. Will also help to deliver improvements along all uprobe consumers (e.g., https://github.com/DataDog/datadog-agent/pull/31298).

### Additional Notes

Tests on load-test-environment against 7.59:

[Dashboard](https://dddev.datadoghq.com/dashboard/gqu-2wz-asy/usm-performance-evaluation-20?fromUser=true&refresh_mode=paused&tpl_var_base_agent-env%5B0%5D=gj-upa-istio-base&tpl_var_client-service%5B0%5D=golang-k6-client-http&tpl_var_compare_agent-env%5B0%5D=gj-upa-istio-new&tpl_var_http.version%5B0%5D=http%2F1.1&tpl_var_kube_cluster_name%5B0%5D=sp-test-env-1&tpl_var_server-service%5B0%5D=golang-httpbin&tpl_var_tls.library%5B0%5D=istio&from_ts=1732282102120&to_ts=1732283494637&live=false)

Screenshot 2024-11-22 at 16 09 59

Accuracy:
Screenshot 2024-11-22 at 16 10 32

Screenshot 2024-11-22 at 16 10 26

Accuracy seems to have been reduced a bit against 7.59. 

Testing again against latest master-py3 image seems to make the accuracy issues less important:

Screenshot 2024-11-22 at 16 49 17

Screenshot 2024-11-22 at 16 48 49

### Describe how to test/QA your changes

Tested in load-test environment, unit tests also test the behavior.

### CODE REVIEW

File: pkg/network/usm/istio.go [hunk 1]
```diff
@@ -81,64 +80,42 @@ var istioProbes = []manager.ProbesSelector{
...
 
-// DetachPID detaches a given pid from the eBPF program
-func (m *istioMonitor) DetachPID(pid uint32) error {
-	return m.registry.Unregister(pid)
-}
+	attachCfg := uprobes.AttacherConfig{
+		ProcRoot: c.ProcRoot,
+		Rules: []*uprobes.AttachRule{{
+			Targets:          uprobes.AttachToExecutable,
+			ProbesSelector:   nodeJSProbes,
```
Comment (vitkyrka) [inline pkg/network/usm/istio.go:102]: istioProbes

File: pkg/network/usm/istio.go [hunk 2]
```diff
@@ -81,64 +80,42 @@ var istioProbes = []manager.ProbesSelector{
...
+			ExecutableFilter: m.isIstioBinary,
+		}},
+		EbpfConfig:                     &c.Config,
+		ExcludeTargets:                 uprobes.ExcludeSelf | uprobes.ExcludeInternal | uprobes.ExcludeBuildkit | uprobes.ExcludeContainerdTmp,
```
Comment (vitkyrka) [inline pkg/network/usm/istio.go:106]: These exclude flags don't reflect the behaviour of the current code AFAICS?

File: pkg/network/usm/istio.go (hunk 1 above)
Comment (gjulianm) [reply pkg/network/usm/istio.go:102]: woops

File: pkg/network/usm/istio.go (hunk 2 above)
Comment (gjulianm) [reply pkg/network/usm/istio.go:106]: The behavior to block buildkit/containerd is  in the `addHooks` function in `ebpf_ssl.go`: https://github.com/DataDog/datadog-agent/blob/guillermo.julian/uprobe-attacher-istio/pkg/network/usm/ebpf_ssl.go#L612-L616.

Self and Internal are not there, but considering the agent and related processes are not going to match anyways we save some performance by adding this and not inspecting those binaries.

File: pkg/network/usm/istio.go [hunk 3]
```diff
@@ -147,48 +124,7 @@ func (m *istioMonitor) Start() {
...
-			m.wg.Done()
-		}()
-
-		for {
-			select {
-			case <-m.done:
-				return
-			case <-processSync.C:
-				m.sync()
-				m.registry.Log()
```
Comment (vitkyrka) [inline pkg/network/usm/istio.go:186]: We're loosing this debug information with this change. Looks like we lost it for nodejs too. Maybe we can add it to the attacher? Can be in a separate PR.

**Review by vitkyrka — APPROVED** [review]

**Review by guyarb — CHANGES_REQUESTED** [review]
Please don't merge this PR for 7.61
We have too many changes in the istio code base, and it is going to be enabled by default in 7.61
Hence I'd like to limit the changes in the module for this version

File: pkg/network/usm/istio.go [hunk 4]
```diff
@@ -81,64 +80,42 @@ var istioProbes = []manager.ProbesSelector{
...
+		EbpfConfig:                     &c.Config,
+		ExcludeTargets:                 uprobes.ExcludeSelf | uprobes.ExcludeInternal | uprobes.ExcludeBuildkit | uprobes.ExcludeContainerdTmp,
+		EnablePeriodicScanNewProcesses: true,
+	}
 
-var (
-	// ErrNoEnvoyPath is returned when no envoy path is found for a given PID
-	ErrNoEnvoyPath = fmt.Errorf("no envoy path found for PID")
-)
+	m.processMonitor = monitor.GetProcessMonitor()
```
Comment (guyarb) [inline pkg/network/usm/istio.go:110]: I'd move it to the declaration of `m` in line 93

File: pkg/network/usm/istio.go [hunk 5]
```diff
@@ -81,64 +80,42 @@ var istioProbes = []manager.ProbesSelector{
...
 
-// AttachPID attaches a given pid to the eBPF program
-func (m *istioMonitor) AttachPID(pid uint32) error {
-	path := m.getEnvoyPath(pid)
-	if path == "" {
-		return ErrNoEnvoyPath
+	attacher, err := uprobes.NewUprobeAttacher(consts.USMModuleName, istioAttacherName, attachCfg, mgr, nil, &uprobes.NativeBinaryInspector{}, m.processMonitor)
+	if err != nil {
+		log.Errorf("Cannot create uprobe attacher: %v", err)
 	}
```
Comment (guyarb) [inline pkg/network/usm/istio.go:115]: If there's an error, we just log and continue? that does not make sense to me

File: pkg/network/usm/istio.go [hunk 6]
```diff
@@ -81,64 +80,42 @@ var istioProbes = []manager.ProbesSelector{
...
 
-	return m.registry.Register(
-		path,
-		pid,
-		m.registerCB,
-		m.unregisterCB,
-		utils.IgnoreCB,
-	)
+	m.attacher = attacher
+	return m
```
Comment (guyarb) [inline pkg/network/usm/istio.go:118]: I would create `m` only at the end of the function where I have all attributes

File: pkg/network/usm/istio.go [hunk 7]
```diff
@@ -198,62 +134,11 @@ func (m *istioMonitor) Stop() {
...
-func (m *istioMonitor) handleProcessExit(pid uint32) {
-	// We avoid filtering PIDs here because it's cheaper to simply do a registry lookup
-	// instead of fetching a process name in order to determine whether it is an
-	// envoy process or not (which at the very minimum involves syscalls)
-	_ = m.DetachPID(pid)
-}
-
-func (m *istioMonitor) handleProcessExec(pid uint32) {
-	_ = m.AttachPID(pid)
+	m.attacher.Stop()
```
Comment (guyarb) [inline pkg/network/usm/istio.go:137]: Currently, this code piece can panic. 
The constructor continues even if we had issue to generate the attacher, hence, stopping would yield a panic

File: pkg/network/usm/istio.go [hunk 8]
```diff
@@ -147,48 +124,7 @@ func (m *istioMonitor) Start() {
...
-				return
-			case <-processSync.C:
-				m.sync()
-				m.registry.Log()
-			}
-		}
-	}()
-
-	utils.AddAttacher(consts.USMModuleName, "istio", m)
+	_ = m.attacher.Start()
```
Comment (guyarb) [inline pkg/network/usm/istio.go:127]: why not logging in case of an error?

File: pkg/network/usm/istio.go (hunk 4 above)
Comment (gjulianm) [reply pkg/network/usm/istio.go:110]: Done

Comment (gjulianm) [reply pkg/network/usm/istio.go:110]: Actually, I ended up moving it down below after the attacher creation, so that we only create it if the attacher hasn't had an error.

File: pkg/network/usm/istio.go (hunk 5 above)
Comment (gjulianm) [reply pkg/network/usm/istio.go:115]: Changed the function signature to return an error.

File: pkg/network/usm/istio.go (hunk 6 above)
Comment (gjulianm) [reply pkg/network/usm/istio.go:118]: The main problem is that the attacher config requires the executable filter to be `m.isIstioBinary` (that way we can pass the config value for the envoy cmd) so we need to create the `istioMonitor` before.

File: pkg/network/usm/istio.go (hunk 7 above)
Comment (gjulianm) [reply pkg/network/usm/istio.go:137]: Fixed

File: pkg/network/usm/istio.go (hunk 8 above)
Comment (gjulianm) [reply pkg/network/usm/istio.go:127]: Fixed

File: pkg/network/usm/istio.go [hunk 9]
```diff
@@ -124,7 +124,14 @@ func (m *istioMonitor) Start() {
...
 	}
 
-	_ = m.attacher.Start()
+	if m.attacher == nil {
+		log.Errorf("Istio monitoring is enabled but the attacher is nil")
+	}
+
+	if err := m.attacher.Start(); err != nil {
+		log.Errorf("Cannot start istio attacher: %s", err)
+	}
```
Comment (guyarb) [inline pkg/network/usm/istio.go:133]: if `m.attacher == nil` you'll panic in line 131

File: pkg/network/usm/istio.go [hunk 10]
```diff
@@ -124,7 +124,14 @@ func (m *istioMonitor) Start() {
 		return
 	}
 
-	_ = m.attacher.Start()
+	if m.attacher == nil {
+		log.Errorf("Istio monitoring is enabled but the attacher is nil")
```
Comment (guyarb) [inline pkg/network/usm/istio.go:128]:
```suggestion
		log.Error("istio monitoring is enabled but the attacher is nil")
```

File: pkg/network/usm/istio.go [hunk 11]
```diff
@@ -125,14 +125,15 @@ func (m *istioMonitor) Start() {
 	}
 
 	if m.attacher == nil {
-		log.Errorf("Istio monitoring is enabled but the attacher is nil")
+		log.Error("istio monitoring is enabled but the attacher is nil")
+		return
```
Comment (guyarb) [inline pkg/network/usm/istio.go:129]: in the `Stop` method you should be wrapping `m.attacher.Stop` with `if m.attacher != nil`

**Review by guyarb — APPROVED** [review]
